File: src/main/java/org/projectfk/blog/data/Blog.java
package org.projectfk.blog.data;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import org.jetbrains.annotations.NotNull;
import org.projectfk.blog.common.NotFoundException;
import org.projectfk.blog.services.BlogService;

import javax.persistence.*;
import java.time.LocalDateTime;

import static com.fasterxml.jackson.annotation.JsonProperty.Access.READ_ONLY;

@JsonIgnoreProperties(ignoreUnknown = true)
@Entity(name = "Blog")
public class Blog {

	//	JPA
	protected Blog() {
	}

	public Blog(
			@NotNull
					User author,
			@NotNull
					String title,
			@NotNull
                    String content,
            @NotNull
                    Tag tag
	) {
		this.author = author;
		this.title = title;
		this.content = content;
        this.tag = tag;
	}

	@JsonProperty(access = READ_ONLY)
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private int id = 0;

	@JsonProperty(access = READ_ONLY)
	@JoinColumn(name = "user")
	@ManyToOne
	private User author;

	@JsonProperty
	@Column(nullable = false)
	private String title;

	@JsonProperty
	@Column(nullable = false)
	private String content;

	@JsonProperty(access = READ_ONLY)
	@UpdateTimestamp
	private LocalDateTime modifyDate = LocalDateTime.now();

	@JsonProperty(access = READ_ONLY)
	@CreationTimestamp
	private LocalDateTime createdDate = LocalDateTime.now();

    @JsonProperty
    @Enumerated(EnumType.STRING)
    private Tag tag = Tag.tech;

	public int getId() {
		return id;
	}

	public User getAuthor() {
		return author;
	}

	public void setAuthor(User author) {
		this.author = author;
	}

	public String getTitle() {
		return title;
	}

	public void setTitle(String title) {
		String old = this.title;
		this.title = title;
	}

	public String getContent() {
		return content;
	}

	public void setContent(String content) {
		String old = this.content;
		this.content = content;
	}

	public boolean alreadyLoaded() {
		return id == 0;
	}

	public LocalDateTime getModifyDate() {
		return modifyDate;
	}

	public LocalDateTime getCreatedDate() {
		return createdDate;
	}

	public Tag getTag() {
		return tag;
	}

	public void setTag(Tag tag) {
		this.tag = tag;
	}

	public static Blog jsonIDEntry(int id) throws NotFoundException {
		return BlogService
				.getService()
				.blogByID(id)
				.orElseThrow(() -> new NotFoundException("There's no blog with id:" + id + " in database"));
	}

	@Override
	public String toString() {
		return "Blog{" +
				"id=" + id +
				", author=" + author +
				", title='" + title + '\'' +
				", content='" + content + '\'' +
				", modifyDate=" + modifyDate +
				", createdDate=" + createdDate +
				'}';
	}
}

File: src/main/java/org/projectfk/blog/data/Blog.kt
package org.projectfk.blog.data

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonProperty.Access.READ_ONLY
import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty
import org.projectfk.blog.blogformat.BlogFormat
import java.io.Serializable
import java.time.LocalDateTime
import javax.persistence.*

@Entity(name = "Blog")
class Blog(postName: String, content: String,
           author: User,
           format: BlogFormat) : Serializable {

//   JPA
    @JsonCreator(mode = JsonCreator.Mode.DISABLED)
    constructor() : this("", "", User(), BlogFormat.PLAIN_TEXT)

    @Id
    @Column(nullable = false)
    @GeneratedValue(strategy = GenerationType.IDENTITY)

    @JacksonXmlProperty(isAttribute = true)
    @JsonProperty(access = READ_ONLY)
    val id: Int = 0



    @Column(nullable = false)

    @JsonProperty
    val author = author


    @Column(nullable = false)
    @Enumerated(EnumType.STRING)

    @JsonProperty
    val format = format


    @Column(nullable = false)

    @JsonProperty("title")
    var postTitle: String = postName
        set(value) {
            field = value
            refreshModifyDate()
        }


    @Column(nullable = false)

    @JsonProperty
    var rawContent: String = content
        set(value) {
            field = value
            refreshModifyDate()
        }


    @Column(nullable = false)

    @JsonProperty(access = READ_ONLY)
    val createDate: LocalDateTime = LocalDateTime.now()


    @Column(nullable = false)

//    JsonIgnore performs buggy on kotlin
    @JsonProperty(access = READ_ONLY)
    var modifyDate: LocalDateTime = LocalDateTime.now()
        private set

    private fun refreshModifyDate() {
        modifyDate = LocalDateTime.now()
    }

    override fun toString(): String =
            "Blog(author=$author, format=$format, id=$id, postTitle='$postTitle', rawContentLength='${rawContent.length}', createDate=$createDate, modifyDate=$modifyDate)"

    companion object {

        @JsonCreator
        @JvmStatic
        fun jsonEntry(
                @JsonProperty("id", required = true)
                id: Int): Blog {
            TODO()
        }

    }

}
